### PR TITLE
Share the implementation of hashing for both vars_prompt and password_hash

### LINKED ISSUE #15326: password_hash/get_encrypted_password uses passlib default of rounds=656000 which is 131 times glibc default
In the password_hash filter function the underlying passlib call uses the default rounds parameter. 

The default for glibc is 5000, the passlib default for sha512 is 656000. This means on a login in a linux account the hash calculation will take significantly longer.

Actually you get basically no rounds parameter when setting it to 5000

``` shell
$ python -c "from passlib.hash import sha512_crypt; print sha512_crypt.encrypt('foo', rounds=5000)"
$6$0zDM3P3/MbIoF0G0$pVmR8hc0ZNYdfGLnFhVKeCGhd32kDXo6ky83JmUWTZCfVQm2IrhnIbrrg5Vyl9XxE5aWzdmuYpvTA6gyOFc5Z.
$ python -c "from passlib.hash import sha512_crypt; print sha512_crypt.encrypt('foo', rounds=5001)"
$6$rounds=5001$VFyJ36YLb/RLhx0e$CAFCB/W7ebYEHIFsZtlJSdvuzEtsYOAvtOwCF7ahqpWqLj62TJp4LlzZ/FiW1B2U5kSKh4xibiJgZyd2ceVoI1
```

Could a simple parameter be added to get_encrypted_password to set the value which has a preferable default what glibc does?

### LINKED ISSUE #17266: properly support passlib module in password_hash
##### ISSUE TYPE
- Feature Idea
- Documentation Report
##### COMPONENT NAME

`password_hash`
##### ANSIBLE VERSION

2.1.1.0-1
##### SUMMARY

The `password_hash` filter has been updated to support the `passlib` python module (as can be seen here: https://github.com/ansible/ansible/blob/v2.1.1.0-1/lib/ansible/plugins/filter/core.py#L279).
However:
1. [The documentation](https://docs.ansible.com/ansible/playbooks_filters.html#hashing-filters) still says it depends on `crypt`.
2. The filter restricts usage to md5, blowfish, sha256, and sha512. The `passlib` module supports a TON more hashing schemes than this. The restriction only made sense when using the `crypt` module. It should not be applied to `passlib`.

### PR DESCRIPTION
**Pull request by mfuchs**
##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
vars_prompt
filter

##### ANSIBLE VERSION
```
2.2.1
```

##### SUMMARY

Shares the implementation of hashing secrets for both vars_prompt and password_hash.

* vars_prompt with encrypt does not require passlib for the algorithms
  supported by crypt.
* Additional checks ensure that there is always a result.
  This works around issues in the crypt.crypt python function that returns
  None for algorithms it does not know.
  Some modules (like user module) interprets None as no password at all,
  which is misleading.
* The password_hash filter supports all parameters of passlib.
  This allows users to provide a rounds parameter, fixing #15326.
* password_hash is not restricted to the subset provided by crypt.crypt,
  fixing one half of #17266.
* Updated documentation fixes other half of #17266.
* password_hash does not hard-code the salt-length, which fixes bcrypt
  in connection with passlib.
  bcrypt requires a salt with length 22.
* Salts are only generated by ansible when using crypt.crypt.
  Otherwise passlib generates them.
* Avoids deprecated functionality of passlib with newer library versions.

Fixes #15326
Fixes #17266

```
# after
{{ 'secret' | password_hash('bcrypt', rounds=14) }}
```

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to unit test errors. Here's one of the errors:

```
2017-02-09 23:40:23 =================================== FAILURES ===================================
2017-02-09 23:40:23 _________________________ TestEncrypt.test_do_encrypt __________________________
2017-02-09 23:40:23 
2017-02-09 23:40:23 self = <units.utils.test_encrypt.TestEncrypt testMethod=test_do_encrypt>
2017-02-09 23:40:23 
2017-02-09 23:40:23     def test_do_encrypt(self):
2017-02-09 23:40:23         self.assertTrue(encrypt.PASSLIB_AVAILABLE)
2017-02-09 23:40:23     
2017-02-09 23:40:23 >       with self.assertRaises(AnsibleError):
2017-02-09 23:40:23 E       TypeError: failUnlessRaises() takes at least 3 arguments (2 given)
2017-02-09 23:40:23 
2017-02-09 23:40:23 test/units/utils/test_encrypt.py:76: TypeError
```

**Comment by mattclay** [comment]
CI failure in unit tests due to:

```
2017-02-10 21:37:00 E           MissingBackendError: bcrypt: no backends available -- recommend you install one (e.g. 'pip install bcrypt')
```

The full error is quite long, so I'll just link to the logs here: https://app.shippable.com/runs/589e2dee031d631000bae9d4/2/console

**Comment by mfuchs** [comment]
Thanks mattclay for your input! :)

ready_for_review

**Comment by abadger** [comment]
Catch and reraise is a better pattern than passing the exception in.  Python 2 makes this a pain, though.    Ned Batchelder has a good blog post about how to do that right: https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html

passlib changes number of rounds while crypt has a static value.  I'm not sure of al the ramifications for that.  It would mean that pre and post change, lookup('password_hash'[...]) will return different strings.  If rounds is calculated dynamically at runtime, it could mean that when I use a playbook that has lookup('password_hash'[...]) on my low end laptop, I generate a different string than when someone else runs it on their high end workstation.  If that's true it leads to non-repeatability of playbook runs.

**Comment by mfuchs** [comment]
As discussed with abadger I will make sure that password_hash returns the same value for both 2.2 and when this commit is merged.
No matter if passlib is installed or not.
This will also be part of the unit test.

Furthermore I will look into allowing a rounds parameter for crypt, for example crypt.crypt('secret', '$6$rounds=656000$abc').

Maybe I will also update the vars_prompt documentation to mention that it makes sense to specify a rounds parameter to ensure replayability.

**Review by dharmabumstead — APPROVED** [review]
Doc updates are OK. Thanks @mfuchs!

**Comment by mfuchs** [comment]
Finally had time to work on this again.

Further changes:
* Always use the crypt rounds defaults when no rounds were specified manually. This will lead to new hashes for vars_prompt. They will be idempotent from then on.
* Fixes #25347 except bcrypt still uses 2a, instead of the suggested 2b.

To discuss:
* what about vars_prompt? Should there be a way to specify the rounds?
* @dharmabumstead I can't use the nicer tests as not every machine has the needed libraries installed. That is why I have these ugly checks for passlib and darwin. I'd be very grateful if you have a better idea, because this is really ugly. :(

**Comment by dharmabumstead** [comment]
Closing due to inactivity. Please feel free to reopen and rebase.

**Comment by Akasurde** [comment]
@The-Wabbit @pilou- @bcoca Could you please review this ?

**Comment by BarbzYHOOL** [comment]
Every commands takes 15sec until this is merged, this is crazy. Users passwords are set with that horrible setting "rounds=656000"

ref: https://serverfault.com/questions/909248/can-you-set-the-number-of-rounds-that-ansibles-password-hash-jinja2-filter-uses

File: lib/ansible/plugins/filter/core.py
Comment (pilou-) [inline lib/ansible/plugins/filter/core.py]: Should not `None` be returned when `hashtype` is unknown instead of raising `AnsibleFilterError` ? I mean without this change, `None` is returned when `hashype` is unknown, is there a reason to change that ?

File: lib/ansible/utils/encrypt.py
Comment (pilou-) [inline lib/ansible/utils/encrypt.py]: Should not `isinstance(self.crypt_algo, HasRawSalt)` be used instead of accessing a [private helper](https://bitbucket.org/ecollins/passlib/src/849ab1e6b5d4ace4c727a63d4adec928d6d72c13/passlib/utils/handlers.py?at=default&fileviewer=file-view-default#handlers.py-1236) ?

File: lib/ansible/utils/encrypt.py
Comment (pilou-) [inline lib/ansible/utils/encrypt.py]: Error message should contain `passlib.__version__`

File: lib/ansible/utils/encrypt.py
Comment (pilou-) [inline lib/ansible/utils/encrypt.py]: `crypt_hash` and `passlib_hash` methods don't seem useful: `passlib_or_crypt` could use `CryptHash`/`PasslibHash` directly.

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: New unit tests must be `pytest` based (not `unittest` based).

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: With shippable, `IS_DARWIN` will always be `False` (unit tests aren't executed using OS X).

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: Should not `passlib` be available ?

File: lib/ansible/plugins/filter/core.py
Comment (mfuchs) [reply lib/ansible/plugins/filter/core.py]: Because I think otherwise there could be unexpected behaviour. A typo in the hash-type or a unsupported type by crypt should not lead to the function being successful. Instead the issue should be pointed out right away.
This also makes the function more consistent with the old do_encrypt.py of encrypt.py.

File: lib/ansible/utils/encrypt.py
Comment (mfuchs) [reply lib/ansible/utils/encrypt.py]: That is a good idea.
I just tried and there seem to be issues:
```python
    assert type(passlib.hash.bcrypt) is abc.ABCMeta
```
It appears to me that some python magic is done in passlib/ifc.py with PasswordHash which seems to kinda break your approach. :(

**Comment by mfuchs** [comment]
@pilou- thanks for your comments! I'll check through them and then will push a cleaned-up and rebased version the next days. :)

**Comment by BarbzYHOOL** [comment]
Awesome, I hope this gets merged

File: test/units/utils/test_encrypt.py
Comment (mfuchs) [reply test/units/utils/test_encrypt.py]: At least last time the tests were run passlib was not available on all test-platforms.

**Comment by mfuchs** [comment]
@pilou- can you please re-review my changes? :)

File: lib/ansible/utils/encrypt.py
Comment (pilou-) [reply lib/ansible/utils/encrypt.py]: Indeed, `issubclass` should be used:
```
In [1]: issubclass(getattr(passlib.hash, 'sha256_crypt'), HasRawSalt)
Out[1]: False

In [2]: issubclass(getattr(passlib.hash, 'pbkdf2_sha256'), HasRawSalt)
Out[2]: True
```

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: Should not an algorithm using a _raw salf_ be tested ?

File: lib/ansible/utils/encrypt.py
Comment (pilou-) [inline lib/ansible/utils/encrypt.py]: Should not `ansible.utils.encrypt.random_password` be reused here ([example](https://github.com/ansible/ansible/blob/adc0efe10c5be833b5138e1d8b0d0316f87c5241/lib/ansible/plugins/lookup/password.py#L207)) ?

Comment (mfuchs) [reply lib/ansible/utils/encrypt.py]: I did some modifications:
* Added encrypt.random_salt which is called by password._random_salt and also used in encrypt.py
* random_salt does not do a to_text. I am not sure if that is actually needed.
  Before to_text was indirectly used via _gen_candiate_chars

**Comment by pilou-** [comment]
Network failures on Fedora/RHEL/CentOS:
* RHE/7.4/1
```
2018-08-09 15:52:25 TASK [yum : download an rpm] ***************************************************
2018-08-09 15:52:25 fatal: [testhost]: FAILED! => {"changed": false, "dest": "/tmp/fpaste.rpm", "msg": "Request failed", "response": "HTTP Error 404: Not Found", "state": "absent", "status_code": 404, "url": "http://download.fedoraproject.org/pub/epel/7/x86_64/Packages/f/fpaste-0.3.7.4.1-1.el7.noarch.rpm"}
```
* CentOS 6
```
2018-08-09 15:52:56 TASK [yum : download an rpm] ***************************************************
2018-08-09 15:52:56 fatal: [testhost]: FAILED! => {"changed": false, "dest": "/tmp/fpaste.rpm", "msg": "Request failed", "response": "HTTP Error 404: Not Found", "state": "absent", "status_code": 404, "url": "http://download.fedoraproject.org/pub/epel/7/x86_64/Packages/f/fpaste-0.3.7.4.1-1.el7.noarch.rpm"}
```

**Comment by mattclay** [comment]
I've fixed the yum tests and restarted CI for this PR.

File: lib/ansible/plugins/lookup/password.py
Comment (pilou-) [inline lib/ansible/plugins/lookup/password.py]: Could not `_random_salt` method be deleted ?

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: `pytest.skip("passlib not available")` should be used ?

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: Should not this block moved at the beginning of the test method (in order to be executed even when passlib isn't available) ?

**Comment by mfuchs** [comment]
@pilou- thanks for your input! I pushed another commit to make it easier for you to check the diff.

Should I squash them all toghether and rebase them on most recent devel?

File: test/units/utils/test_encrypt.py
Comment (pilou-) [inline test/units/utils/test_encrypt.py]: It seems best to split the `test_do_encrypt`: one part without passlib (first part of the current method), one with (this one will use `pytest.skip`). Sorry, my previous comments were not clear :-/

**Comment by pilou-** [comment]
> Should I squash them all toghether and rebase them on most recent devel?

no need to squash them, there will be squashed when the PR is merged (see [committers guidelines](https://docs.ansible.com/ansible/latest/community/committer_guidelines.html?highlight=squash#general-rules))

File: test/units/utils/test_encrypt.py
Comment (mfuchs) [reply test/units/utils/test_encrypt.py]: Indeed. I realised that myself just when I left for lunch. :)

**Review by pilou- — APPROVED** [review]
shipit

File: lib/ansible/plugins/filter/core.py
Comment (abadger) [inline lib/ansible/plugins/filter/core.py]: I'm not a fan of passing opaque data structures (```settings```) around.  I'd much rather that these parameters were explicitly listed. (In the other functions as well).

File: lib/ansible/plugins/filter/core.py
Comment (abadger) [reply lib/ansible/plugins/filter/core.py]: Typically, our code uses try: except to catch and transform exception rather than making explicit passing of exception types part of a function's signature.  There are drawbacks to try: except (for instance, changing the traceback on python2) but I think we should stick with that idiom.

File: lib/ansible/utils/encrypt.py
Comment (abadger) [inline lib/ansible/utils/encrypt.py]: Don't use the name _encrypt for hashing... I know we already have a publically available function named do_encrypt which really does hashing but there's no sense making that problem worse.

File: lib/ansible/utils/encrypt.py
Comment (abadger) [inline lib/ansible/utils/encrypt.py]: Note: organizationally, our code is largely, callers near the bottom of the file, callees near the top.  (Like C code).

File: lib/ansible/utils/encrypt.py
Comment (abadger) [inline lib/ansible/utils/encrypt.py]: For _clean_salt and _clean_rounds, just pass in the one setting you care about and then return the modified version of it.  Change the container in the caller.  functional programming like that is easier to test and easier to reason about.

**Comment by abadger** [comment]
Okay, I took a look.  the logic here looks fine but there's some coding style/API that I think should be changed.  I'll be busy with the new release so if you get some time to work on this but I don't notice, feel free to ping me on IRC (abadger1999 on irc.freenode.net, the #ansible-devel channel).

File: lib/ansible/plugins/filter/core.py
Comment (mfuchs) [reply lib/ansible/plugins/filter/core.py]: I modified it to always raise AnsibleError.
So also in the filter context (password_hash).
Please tell me whether I should raise AnsibleFilterError there (via the steps you suggested -> https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html)?

**Comment by mfuchs** [comment]
@abadger all comments should be handled now. Thanks for your input!

Just in case the rethrowing with keeping the traceback works differently in Python 2 and Python3.
Thus the code is quite ugly.

If this is a common use case then it should be moved to a better place (a function), something like reraise_as.
Of course reraise_as should then modify the traceback to not be part of it.

**Comment by abadger** [comment]
For the raise syntax issues... we do bundle six as ansible.module_utils.six... six has a couple raise functions but I'm not sure if they're precisely what you need or not.  (Also, if one of them can do what ned Batchelder's recipe does, sorry for not thinking of it earlier... I'll blame lack of sleep for not remembering those if so ;-)

I'll look in on this in the morning.

**Comment by mfuchs** [comment]
Some tests fail.
I checked and at least test/integration/targets/vars_prompt/vars_prompt-5.yml fails.
The reason is that it expects `re.compile('"password": "\\$6\\$rounds=')` in the password-output.
Yet since no rounds are provided the default of crypt is used leading to a output like

> "password": "$6$jESIyad4F08hP3Ta$hrnwEbDk9uSg1IURdckNLBy9xoeO1y1peRDBHsdmwzlWv1u9Cs0qasiy7THQ88dnzC0f4PaFX4ySUCcyT5emn0"

This change was implemented to be more consistent overall, thus having consistency between vars_prompt and password_hash (no matter if passlib is installed or not) and not depending on the always changing defaults of passlib.

**Comment by abadger** [comment]
Makes sense to me.  Go ahead and change the test.

**Comment by mfuchs** [comment]
Since the test is rather new I rebased on current devel + adapted the integration test.

**Comment by abadger** [comment]
Merged to devel for the 2.7.0 release.  Thanks @mfuchs

**Comment by BarbzYHOOL** [comment]
awesome

**Pull request merged**
